### ISSUE TITLE
tsconfig 'path mapping' not working with wallaby in Sublime Text 3 

### ISSUE THREAD

**Issue by joelhoward0** [wallabyjs/public#1082]
### Issue description or question

Since updating my project to use Typescript 2's Path Mapping feature, wallaby reports 'cannot find modules' errors suggesting that the 'paths' part of my tsconfig isn't being used properly (or something like that). In the error below, 'hex' module is on physical path /domain/hex/hex, aliases to /hex/hex in the tsconfig.

Note that if I change the path to 'domain/hex/hex', it works - so 'baseUrl' seems to be working, but not the mapped paths. Details below, thanks!

Wallaby TS version:
```
console.log('Wallaby ts version:', require('typescript').version); 
prints Wallaby ts version: 2.2.1
```
```
[Error]: Runtime error: Error: Cannot find module 'hex/hex'​​
​​[Error]:     at Function.Module._resolveFilename (module.js:469:15)​​
​​[Error]:     at Function.Module._load (module.js:417:25)​​
​​[Error]:     at Module.require (module.js:497:17)​​
​​[Error]:     at require (internal/module.js:20:19)​​
​​[Error]:     at Object.<anonymous> (.\tests\hex-tests.js:6:28)​​
```

the relevant line in the test file:
```
import { Hex } from 'hex/hex'`
```
### tsconfig.json configuration file

```
{
  "compilerOptions": {
    "module": "commonjs",
    "sourceMap": true,
    "outDir": "./compiledts",
    "target": "es6",
    "baseUrl": ".",
    "paths": {
      "routes/*": ["client/components/routes/*"],
      "components/*": ["client/components/*"],
      "hex/*": ["domain/hex/*"]
    }
  },
  "exclude": [
      "node_modules",
      "typings/browser",
      "typings/browser.d.ts"
  ],
  "compileOnSave": false,
  "atom": {
      "rewriteTsconfig": false
  }
}
```

### Wallaby.js configuration file
```
module.exports = function(wallaby) {
  return {
    files: [
      'domain/hex/*.ts'
    ],
    tests: [
      'tests/hex-tests.ts'
    ],
    env: {
      type: 'node'
    },
    workers: {
      initial: 1,
      regular: 1
    }
  }
}
```

### Code editor or IDE name and version
Sublime Text v3

### OS name and version
Windows

**Comment by ArtemGovorov** [comment]
TypeScript compilation [doesn't change the the import source](https://github.com/Microsoft/TypeScript/issues/5039#issuecomment-206451221). The path mapping feature is just really a dev time thing, so one needs to do the runtime mapping depending on the platform. In node.js, one way to do would be to use the `NODE_PATH` env variable to give node.js a hint how to resolve the `hex/hex` like paths.

```javascript
module.exports = function(wallaby) {
  var path = require('path');
  process.env.NODE_PATH += 
    path.delimiter + path.join(wallaby.localProjectDir, 'client/components') +
    path.delimiter + path.join(wallaby.localProjectDir, 'domain');

  return {
    ...
  }
}
```

**Comment by joelhoward0** [comment]
Thanks for the help! I had to tweak your supplied code to reference the wallaby project cache, instead of the local project directory. My version of what you supplied looked like this:

```
module.exports = function(wallaby) {
  var path = require('path');

  process.env.NODE_PATH += 
    path.delimiter + path.join(wallaby.projectCacheDir, 'client/components') +
    path.delimiter + path.join(wallaby.projectCacheDir, 'domain');
```

**Comment by ArtemGovorov** [comment]
Awesome, thanks for the update! Have update my answer to use `projectCacheDir`, sorry about forgetting about it initially.

**Comment by myflowpl** [comment]
I just run on the same issue, and find out more stable solution
install `tsconfig-paths` package in your project
create new file `wallaby-paths.js` next to your `wallaby.js` and `tsconfig.json` files
```javascript
// ./wallaby-paths.js
const tsConfigPaths = require('tsconfig-paths');
const tsconfig = require('./tsconfig.json');
tsConfigPaths.register({
  baseUrl: tsconfig.compilerOptions.baseUrl,
  paths: tsconfig.compilerOptions.paths
});
```
 add `env` config to your `wallaby.js` file
```javascript
// ./wallaby.js 
...
    env: {
      type: 'node',
      params: {
        runner: '-r ' + path.join(__dirname, './wallaby-paths.js')
      }
    },
...
```
this way you manage your `paths` only in one place, inside `tsconfig.json`, and you don't mess with NODE_PATHS directly

BTW
@ArtemGovorov use of tsconfig paths should be build into wallaby.js, or at least the manual configuration should be added to official documentation, be course custom paths are becoming more and more used.
But, anyway, Wallaby ROCKS :1st_place_medal:

**Comment by ArtemGovorov** [comment]
@myflowpl Awesome, thanks for the post. We have added the docs section [describing various ways to handle the scenario](https://wallabyjs.com/docs/integration/typescript.html#typescript-compileroptionspaths-setting). 

BTW, I think you can implement your idea without an additional file, [just with the `setup` function](https://wallabyjs.com/docs/integration/typescript.html#node-environment-with-mochajasmineavatape).

**Comment by myflowpl** [comment]
@ArtemGovorov it's awesome, really clear solution, thanks

**Comment by pndewit** [comment]
The above solutions didn't seem to work for me in combination with either [this](https://github.com/wallabyjs/public/issues/784#issuecomment-246247998) workaround that I have implemented or with my setup. After some struggling I came up with this solution, hope it will help anyone:

    setup: (wallaby => {
        wallaby.delayStart();

        requirejs.config({
            // This part is different from the original workaround
            paths: {
                // paths
            },
            packages: [
                // packages
            ]
        });

        require(['test/unit/setup.js'].concat(wallaby.tests), () => {
            wallaby.start();
        });
    })
        .toString()
        .replace('// packages', aureliaJson.build.bundles.reduce((depList, bundle) => {
            return depList + (bundle.dependencies || []).reduce((prev, curr) => {
                let moduleName, modulePath, moduleMain;
                if(curr.path) {
                    moduleName = moduleMain = curr.name;
                    modulePath = path.relative(
                        __dirname,
                        path.resolve(__dirname, 'aurelia_project', curr.path))
                        .split('\\').join('/');
                    if(curr.main) {
                        moduleMain = curr.main;
                    }
                } else {
                    moduleName = moduleMain = curr;
                    modulePath = 'node_modules/' + moduleName + '/dist/amd';
                }
                return prev
                    + '{ name: ' + JSON.stringify(moduleName)
                    + ', location: ' + JSON.stringify(modulePath)
                    + ', main: ' + JSON.stringify(moduleMain)
                    + '},';
            }, '');
        }, ''))

        // This part is different from the original workaround
        .replace('// paths', Object.keys(aureliaJson.paths).reduce((pathList, path) => {
            return pathList + path + ': ' + JSON.stringify(aureliaJson.paths[path]) + ',';
        }, ''))

In short I added 2 new pieces (highlighted with a comment):

    paths: {
        // paths
    },

This is used to be replaced by the second piece (like the original workaround does with the packages).

    .replace('// paths', Object.keys(aureliaJson.paths).reduce((pathList, path) => {
        return pathList + path + ': ' + JSON.stringify(aureliaJson.paths[path]) + ',';
    }, ''))

This adds the paths (aliases) inside my `aurelia_project/aurelia.json` to the requirejs config. I am now also able to map unknown paths to known paths (e.g. mapping: `"root"` to `"src/app"` which is added by the Aurelia CLI by default).
